ChunkRef, tcp_nodelay and UDP max_packet_size

adamhass opened the pull request:
Closes #100, #74 , #96 

udp MAX_PACKET_SIZE compares chunk_size to determine what the maximum amount of receivable bytes is on the receiving end. Users really shouldn't use chunk_size smaller than UDP max packet size.

tcp_nodelay network_config parameter added, allows Kompact users to select, by default nodelay is OFF such that congestion is minimized and throughput maximized.

chunk_ref structure added. A ChunkLease may be converted into a ChunkRef which can then be cloned. To combine ChunkRefs the resulting ChunkRef must be created **from a ChunkLease and a ChunkRef** using the ChunkLease methods `into_chunk_ref_with_head(self, head: ChunkRef)` and `into_chunk_ref_with_tail(self, tail: ChunkRef)`

Bathtor commented on core/src/dispatch/mod.rs:
```diff
@@ -125,6 +129,18 @@ impl NetworkConfig {
+        self.tcp_nodelay
+    }
+
+    /// If set to true the Nagle algorithm will be turned off for all TCP Network-channels
```
Leave a newline after the first line.

Write `false` (with backticks) instead of FALSE

Bathtor commented on core/src/net/buffers/chunk_ref.rs:
```diff
@@ -0,0 +1,233 @@
+use crate::prelude::Buf;
+use std::sync::Arc;
+
+/// A ChunkRef is a Byte-buffer which locks the BufferChunk which the ChunkRef is a lease from.
```
Please use links or code inserts when referring to type names, etc

Bathtor commented on core/src/net/buffers/chunk_ref.rs:
```diff
@@ -0,0 +1,233 @@
+    }
+
+    /// The full length of the underlying bytes without read/write-pointers
+    /// Use `remaining()`/`remaining_mut()` for readable/writable lengths
```
Use links please

Bathtor commented on core/src/net/buffers/chunk_ref.rs:
```diff
@@ -0,0 +1,233 @@
+
+    /// Use different data sizes and buffer sizes to test different kinds of chains/splits.
+    /// We want to test Chaining, Creation functions from chunk_lease.rs and reading/using Buf impl
+    /// Variants:
```
I don't think it makes sense to use doc comments for test functions. Should probably just use normal comments

Bathtor commented on core/src/net/udp_state.rs:
```diff
@@ -27,12 +28,20 @@ impl UdpState {
     ) -> Self {
+        // If chunk_size is smaller than MAX_PACKET_SIZE we will use that size as the limit instead.
+        let chunk_size = network_config.get_buffer_config().chunk_size;
+        let max_packet_size = if chunk_size < MAX_PACKET_SIZE {
```
Use the `min` function?

Bathtor commented on core/src/net/udp_state.rs:
```diff
@@ -44,31 +53,41 @@ impl UdpState {
-                    self.outbound_queue.push_front((addr, frame));
-                    interrupts += 1;
-                    if interrupts >= MAX_INTERRUPTS {
+            if frame.len() > self.max_packet_size {
```
Hmm, is this right? Couldn't it happen that some actor has a different chunk size and already serialised it in a contiguous manner? Shouldn't it be the make_contiguous call below that fails if we don't give it a large enough buffer or something?

adamhass replied:
The BufferConfig which (maybe) determines max_packet_size is the one used for the NetworkThread, so I'm assuming the remote systems NetworkConfig is set-up with the same BufferConfig. The line imposes that we won't send packages larger than we can receive. I could remove that check entirely as the udp send or receive will trigger the error/warning instead if the frame is too big.

Bathtor replied:
Yeah, I don't think we should make this assumption. Let it send whatever it can, and let it fail on the receiver side if necessary. I think we should assume that Kompact systems in general won't share the same configuration.

Bathtor replied:
But you do need to make sure that non-contiguous data can be made contiguous on the sending side. I think that's the only check we should have, since we definitely should not send broken packages!

adamhass commented:
I just added even more to this PR. When merged this will also close #97 .

I went with the naming `tell_preserialised(ChunkRef, self)`. To preserialise something the method call from an actor will look like `self.ctx.preserialise(data) -> ChunkRef`.

Preserialisation packages data and serialisation ID only. tell_preserialised adds frameheader, src, dst.

I refactored a bunch of the tests for this to separate out some things, imo it's cleaner but still a bit of a mess. Basically I wanted to move all the tests that may be regarded as pure integration testing (only asserts on actors in the tests) for the networking. They are in the tests folder (dispatch_integration_tests.rs). The *actors* used in those tests, PingerAct/PongerAct/BingPingerAct/BigPongerAct are in the module net_test_helpers in the net/mod.rs file, public for anyone to use if they wish to try to build their own networking layer/network dispatcher/use in other tests etc. I hope this makes sense and that the naming is useful :)

Bathtor commented:
Can you fix [this TODO](https://github.com/adamhass/kompact/blob/24edbaf2c3115a07014a9f7389ade910de76ff3b/core/src/messaging/net_message.rs#L464) as well in this PR, please? @adamhass
